webhook/workspace/handler/immutable.go

```diff
@@ -160,24 +150,42 @@ func (h *WebhookHandler) checkRestrictedAccessAnnotation(req admission.Request)
 	} else {
 		err = h.Decoder.DecodeRaw(req.Object, obj)
 	}
+	if err != nil {
+		return false, err
+	}
 	annotations := obj.GetAnnotations()
 	return annotations[constants.WorkspaceRestrictedAccessAnnotation] == "true", nil
 }
 
-func changePermitted(oldObj, newObj runtime.Object) bool {
+func checkRestrictedWorkspaceMetadata(oldMeta, newMeta *metav1.ObjectMeta) (allowed bool, msg string) {
+	if oldMeta.Annotations[constants.WorkspaceRestrictedAccessAnnotation] == "true" &&
+		newMeta.Annotations[constants.WorkspaceRestrictedAccessAnnotation] != "true" {
+		return false, "cannot disable restricted-access once it is set"
+	}
+	return true, "permitted change to workspace"
+}
+
+func changePermitted(oldObj, newObj runtime.Object) (allowed bool, msg string) {
 	oldCopy := oldObj.DeepCopyObject()
 	newCopy := newObj.DeepCopyObject()
 	oldMeta, ok := oldCopy.(metav1.Object)
 	if !ok {
-		log.Error(fmt.Errorf("Object %s is not a valid k8s object: does not have metadata", oldObj.GetObjectKind()), "Failed to compare objects")
-		return false
+		log.Error(fmt.Errorf("object %s is not a valid k8s object: does not have metadata", oldObj.GetObjectKind()), "Failed to compare objects")
+		return false, "Internal error"
 	}
 	newMeta, ok := newCopy.(metav1.Object)
 	if !ok {
-		log.Error(fmt.Errorf("Object %s is not a valid k8s object: does not have metadata", newObj.GetObjectKind()), "Failed to compare objects")
-		return false
+		log.Error(fmt.Errorf("object %s is not a valid k8s object: does not have metadata", newObj.GetObjectKind()), "Failed to compare objects")
+		return false, "Internal error"
+	}
+	oldLabels, newLabels := oldMeta.GetLabels(), newMeta.GetLabels()
+	if oldLabels[constants.WorkspaceCreatorLabel] != newLabels[constants.WorkspaceCreatorLabel] {
```

Comment:
I assume here we lose a feature that we have - restore workspace creator label, if it's missing.
Just because it's called before the code that do the feature.
But I haven't tested yet.

Reply:
I thought the same, but our code is too complicated for such simple conclusions ;) 

`changePermitted` is only called for other workspace-involved objects, not workspaces themselves. The apply creator label feature applies only to DevWorkspaces themselves, as it's assumed that the DWO will propagate that label to all workspace-related resources it creates.